dask/dataframe/io/parquet/fastparquet.py

```diff
@@ -468,44 +719,60 @@ def read_partition(
                 null_index_name = True
             columns += index
 
-        if pf is None:
-            base, fns = _analyze_paths([piece], fs)
-            scheme = get_file_scheme(fns)
-            pf = ParquetFile(piece, open_with=fs.open)
-            relpath = piece.replace(base, "").lstrip("/")
-            for rg in pf.row_groups:
-                for ch in rg.columns:
-                    ch.file_path = relpath
-            pf.file_scheme = scheme
-            pf.cats = paths_to_cats(fns, scheme)
-            pf.fn = base
-            if null_index_name and "__index_level_0__" in pf.columns:
-                # See "Handling a None-labeled index" comment above
-                index = ["__index_level_0__"]
-                columns += index
-            return pf.to_pandas(columns, categories, index=index)
-        else:
-            if isinstance(pf, tuple):
-                if isinstance(pf[0], list):
-                    pf = _determine_pf_parts(fs, pf[0], pf[1], **kwargs)[1]
-                else:
-                    pf = ParquetFile(
-                        pf[0], open_with=fs.open, sep=fs.sep, **kwargs.get("file", {})
-                    )
-                pf._dtypes = lambda *args: pf.dtypes  # ugly patch, could be fixed
-                pf.fmd.row_groups = None
-            rg_piece = pf.row_groups[piece]
+        # Use global `parquet_file` object.  Need to reattach
+        # the desired row_group
+        parquet_file = kwargs.pop("parquet_file", None)
+
+        if isinstance(piece, tuple):
+            if isinstance(piece[0], str):
+                # We have a path to read from
+                assert parquet_file is None
+                parquet_file = ParquetFile(
+                    piece[0], open_with=fs.open, sep=fs.sep, **kwargs.get("file", {})
+                )
+                rg_indices = piece[1] or list(range(len(parquet_file.row_groups)))
+
+                # `piece[1]` will contain row-group indices
+                row_groups = [parquet_file.row_groups[rg] for rg in rg_indices]
+            elif parquet_file:
+                # `piece[1]` will contain actual row-group objects,
+                # but they may be pickled
+                row_groups = piece[0]
+                if isinstance(row_groups, bytes):
+                    row_groups = pickle.loads(row_groups)
+                parquet_file.fmd.row_groups = row_groups
+                # TODO: Adding `parquet_file._set_attrs()` here seems to
+                # cause a failure in `test_append_with_partition[fastparquet]`
```

Comment:
Related to (this comment)[https://github.com/dask/fastparquet/pull/549#issuecomment-765534918] - I'll look into this, but let me know if you have thoughts @martindurant

Reply:
Not immediately - I guess append updates/mutates the original fmd, but we messed with it in the meantime?

Reply:
Interesting - It seems that there were two issues: (1) Removing `col.meta_data.statistics` was somehow leading to the wrong dtype in the result.  As far as I can tell, it is because the `null_count` is queried somewhere in fastparquet (I didn't confirm this, but found that the error vanished if I kept `statistics.null_count`).  (2) Calling `_set_attrs` reset `pf.cats` - leading to the wrong `Categorical` dtype (so we need to save and reset `cats` after the call).

Reply:
aha, 2) makes sense: the cats are based on the pathnames that the pf sees. Not sure yet about 1).